File: app/src/main/java/br/com/menuvem/lojista/presentation/history/InsumoHistoryScreen.kt
package br.com.menuvem.lojista.presentation.history

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.outlined.Insights
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import br.com.menuvem.lojista.domain.model.HistoricoPreco
import br.com.menuvem.lojista.presentation.components.EmptyState
import br.com.menuvem.lojista.presentation.components.GraficoPreco
import br.com.menuvem.lojista.presentation.components.TendenciaIndicador
import br.com.menuvem.lojista.presentation.components.formatarMoeda
import br.com.menuvem.lojista.ui.theme.PurplePrimary
import br.com.menuvem.lojista.ui.theme.YellowSecondary
import java.time.format.DateTimeFormatter

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun InsumoHistoryScreen(
    insumoId: Long,
    onBack: () -> Unit,
    viewModel: InsumoHistoryViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Column {
                        Text(
                            text = uiState.insumo?.nome ?: "Histórico",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.SemiBold,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                        uiState.insumo?.let {
                            Text(
                                text = "por ${it.unidadeCompra}",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Voltar")
                    }
                },
                actions = {
                    TendenciaIndicador(
                        tendencia = uiState.tendencia,
                        showLabel = true,
                        modifier = Modifier.padding(end = 16.dp)
                    )
                }
            )
        }
    ) { paddingValues ->
        if (uiState.isLoading) {
            Box(
                modifier = Modifier.fillMaxSize().padding(paddingValues),
                contentAlignment = Alignment.Center
            ) { CircularProgressIndicator(color = PurplePrimary) }
        } else {
            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                contentPadding = PaddingValues(bottom = 24.dp)
            ) {
                // Card de resumo
                item {
                    uiState.insumo?.let { insumo ->
                        Card(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(16.dp),
                            colors = CardDefaults.cardColors(
                                containerColor = MaterialTheme.colorScheme.primaryContainer
                            )
                        ) {
                            Row(
                                modifier = Modifier.padding(16.dp),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Column {
                                    Text(
                                        text = "Custo atual",
                                        style = MaterialTheme.typography.labelMedium,
                                        color = MaterialTheme.colorScheme.onPrimaryContainer.copy(alpha = 0.7f)
                                    )
                                    Text(
                                        text = "R$ ${formatarMoeda(insumo.custoAtual)}",
                                        style = MaterialTheme.typography.headlineMedium,
                                        fontWeight = FontWeight.Bold,
                                        color = MaterialTheme.colorScheme.onPrimaryContainer
                                    )
                                    Text(
                                        text = "por ${insumo.unidadeCompra}",
                                        style = MaterialTheme.typography.bodySmall,
                                        color = MaterialTheme.colorScheme.onPrimaryContainer.copy(alpha = 0.7f)
                                    )
                                }
                                Column(horizontalAlignment = Alignment.End) {
                                    Text(
                                        text = "${uiState.historico.size} compras",
                                        style = MaterialTheme.typography.labelSmall,
                                        color = MaterialTheme.colorScheme.onPrimaryContainer.copy(alpha = 0.7f)
                                    )
                                    Text(
                                        text = "R$ ${formatarMoeda(insumo.custoPorUnidadeUso)}/${insumo.unidadeUso}",
                                        style = MaterialTheme.typography.labelMedium,
                                        fontWeight = FontWeight.SemiBold,
                                        color = YellowSecondary
                                    )
                                }
                            }
                        }
                    }
                }

                // Gráfico
                if (uiState.historico.size >= 2) {
                    item {
                        Text(
                            text = "Evolução do preço",
                            style = MaterialTheme.typography.titleSmall,
                            fontWeight = FontWeight.SemiBold,
                            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
                        )
                        ElevatedCard(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(horizontal = 16.dp),
                            elevation = CardDefaults.elevatedCardElevation(2.dp)
                        ) {
                            GraficoPreco(
                                historico = uiState.historico,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(top = 16.dp, bottom = 8.dp)
                            )
                        }
                        Spacer(modifier = Modifier.height(8.dp))
                    }
                }

                // Histório textual
                item {
                    Text(
                        text = "Histórico de compras",
                        style = MaterialTheme.typography.titleSmall,
                        fontWeight = FontWeight.SemiBold,
                        modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
                    )
                }

                if (uiState.historico.isEmpty()) {
                    item {
                        EmptyState(
                            icon = Icons.Outlined.Insights,
                            title = "Sem registros",
                            subtitle = "Este insumo não tem histórico de preços. Finalize uma lista de compras com ele para começar a registrar."
                        )
                    }
                } else {
                    items(uiState.historico.reversed()) { historico ->
                        HistoricoItem(historico = historico)
                    }
                }
            }
        }
    }
}

@Composable
private fun HistoricoItem(historico: HistoricoPreco) {
    val formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy 'às' HH:mm")

    ListItem(
        headlineContent = {
            Text(
                text = "R$ ${formatarMoeda(historico.preco)}",
                style = MaterialTheme.typography.titleSmall,
                fontWeight = FontWeight.SemiBold
            )
        },
        supportingContent = {
            Column {
                Text(
                    text = historico.data.format(formatter),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                if (historico.listaComprasNome.isNotBlank()) {
                    Text(
                        text = historico.listaComprasNome,
                        style = MaterialTheme.typography.bodySmall,
                        color = PurplePrimary.copy(alpha = 0.8f)
                    )
                }
            }
        },
        leadingContent = {
            Icon(
                imageVector = Icons.Default.Receipt,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.5f),
                modifier = Modifier.size(20.dp)
            )
        }
    )
    HorizontalDivider(modifier = Modifier.padding(horizontal = 16.dp))
}

File: lib/presentation/history/insumo_history_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:intl/intl.dart';
import 'package:rxdart/rxdart.dart';

import '../../domain/model/tendencia_preco.dart';
import '../../domain/usecase/tendencia_usecase.dart';
import '../../providers/repository_providers.dart';
import '../../theme/app_theme.dart';
import '../components/back_or_home_button.dart';
import '../components/empty_state.dart';
import '../components/formatters.dart';
import '../components/responsive.dart';
import '../components/tendencia_indicador.dart';

final insumoHistoryProvider = StreamProvider.family((ref, int insumoId) {
  final insumoRepo = ref.watch(insumoRepositoryProvider);
  final historicoRepo = ref.watch(historicoPrecoRepositoryProvider);

  return Rx.combineLatest2(
    insumoRepo.getAllInsumos().map(
      (list) => list.where((i) => i.id == insumoId).firstOrNull,
    ),
    historicoRepo.getHistoricoByInsumo(insumoId),
    (insumo, historico) => (insumo, historico),
  ).asyncMap((tuple) async {
    final (insumo, historico) = tuple;
    TendenciaPreco tendencia = TendenciaPreco.estavel;
    try {
      tendencia = await getTendenciaPreco(historicoRepo, insumoId);
    } catch (_) {}
    return (insumo: insumo, historico: historico, tendencia: tendencia);
  });
});

class InsumoHistoryScreen extends ConsumerWidget {
  const InsumoHistoryScreen({super.key, required this.insumoId});
  final int insumoId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final async = ref.watch(insumoHistoryProvider(insumoId));

    return Scaffold(
      appBar: AppBar(
        leading: const BackOrHomeButton(),
        title: async.maybeWhen(
          data: (d) => Text(d.insumo?.nome ?? 'Histórico'),
          orElse: () => const Text('Histórico'),
        ),
        actions: [
          async.maybeWhen(
            data: (d) => Padding(
              padding: const EdgeInsets.only(right: 16),
              child: TendenciaIndicador(
                tendencia: d.tendencia,
                showLabel: true,
              ),
            ),
            orElse: () => const SizedBox.shrink(),
          ),
        ],
      ),
      body: MaxWidthCenter(
        child: async.when(
          loading: () => const Center(
            child: CircularProgressIndicator(color: purplePrimary),
          ),
          error: (e, _) => Center(child: Text('Erro: $e')),
          data: (d) {
            final insumo = d.insumo;
            final historico = d.historico;
            final formatter = DateFormat("dd/MM/yyyy 'às' HH:mm");
            return ListView(
              padding: const EdgeInsets.only(bottom: 24),
              children: [
                if (insumo != null)
                  Card(
                    margin: const EdgeInsets.all(16),
                    color: Theme.of(context).colorScheme.primaryContainer,
                    child: Padding(
                      padding: const EdgeInsets.all(16),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: [
                          Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Text(
                                'Custo atual',
                                style: Theme.of(context).textTheme.labelMedium,
                              ),
                              Text(
                                'R\$ ${formatarMoeda(insumo.custoAtual)}',
                                style: Theme.of(context)
                                    .textTheme
                                    .headlineMedium
                                    ?.copyWith(fontWeight: FontWeight.bold),
                              ),
                              Text(
                                'por ${insumo.unidadeCompra}',
                                style: Theme.of(context).textTheme.bodySmall,
                              ),
                            ],
                          ),
                          Column(
                            crossAxisAlignment: CrossAxisAlignment.end,
                            children: [
                              Text(
                                '${historico.length} compras',
                                style: Theme.of(context).textTheme.labelSmall,
                              ),
                              Text(
                                'R\$ ${formatarMoeda(insumo.custoPorUnidadeUso)}/${insumo.unidadeUso}',
                                style: Theme.of(context).textTheme.labelMedium
                                    ?.copyWith(
                                      fontWeight: FontWeight.w600,
                                      color: yellowSecondary,
                                    ),
                              ),
                            ],
                          ),
                        ],
                      ),
                    ),
                  ),
                Padding(
                  padding: const EdgeInsets.symmetric(
                    horizontal: 16,
                    vertical: 8,
                  ),
                  child: Text(
                    'Histórico de compras',
                    style: Theme.of(context).textTheme.titleSmall?.copyWith(
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                ),
                if (historico.isEmpty)
                  const EmptyState(
                    icon: Icons.insights_outlined,
                    title: 'Sem registros',
                    subtitle:
                        'Este insumo não tem histórico de preços. Finalize uma lista de compras com ele para começar a registrar.',
                  )
                else
                  ...historico.reversed.map(
                    (h) => Column(
                      children: [
                        ListTile(
                          leading: Icon(
                            Icons.receipt,
                            size: 20,
                            color: Theme.of(context)
                                .colorScheme
                                .onSurfaceVariant
                                .withValues(alpha: 0.5),
                          ),
                          title: Text(
                            'R\$ ${formatarMoeda(h.preco)}',
                            style: Theme.of(context).textTheme.titleSmall
                                ?.copyWith(fontWeight: FontWeight.w600),
                          ),
                          subtitle: Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Text(formatter.format(h.data)),
                              if (h.listaComprasNome.isNotEmpty)
                                Text(
                                  h.listaComprasNome,
                                  style: const TextStyle(color: purplePrimary),
                                ),
                            ],
                          ),
                        ),
                        const Divider(indent: 16, endIndent: 16),
                      ],
                    ),
                  ),
              ],
            );
          },
        ),
      ),
    );
  }
}
